### ISSUE TITLE
TimeoutCancellationException: Timed out waiting for 15000 ms with Phh Superuser

### ISSUE THREAD

**Issue by ildar** [zhanghai/MaterialFiles#842]
Using the [Phh AOSP](https://github.com/phhusson/treble_experimentations/) the Root functionality fails.
I know some changes made in Feb but libsu also fails to work with Phh Superuser.

Some similar fix made in https://github.com/NeoApplications/Neo-Backup/issues/562 (may be helpful)

**Comment by zhanghai** [comment]
I'm using [libsu](https://github.com/topjohnwu/libsu) fopr the root service and I don't think every root app should integrate a fix on their side for this. It would be more reasonable for Phh AOSP to fix their su implementation, or if it's a libsu bug get it fixed in libsu.

**Comment by ildar** [comment]
I just tested the libsu example/test app and it works just fine. The
problem is in the app.
ref: https://github.com/ildar/libsu/actions/runs/2824595588

**Comment by zhanghai** [comment]
I assume the sample app isn't using mount-master, which file managers need to use. Could you double check?

**Comment by ildar** [comment]
That's a little hard for me, I'm not a Java programmer.
The example has several modes (e.g. Sync/Async/...), all working.
Additionally I made an FR at libsu to clarify:
https://github.com/topjohnwu/libsu/issues/129

**Comment by topjohnwu** [comment]
Mount master is root implementation specific, I don't think Phh's SU has mount master support.

**Comment by zhanghai** [comment]
Should libsu (offer the option to) handle the situation that it isn't supported, or should the su impl ignore the argument if it isn't supported?

**Comment by topjohnwu** [comment]
There is no way to really test whether an `su` binary accepts certain arguments. Even if we check its version by calling `su -v`, it's still possible that would hang for certain `su` executables.

**Comment by vvb2060** [comment]
Apps should not rely on su to switch mount namespaces.

open /proc/1/ns/mnt and setns

**Comment by topjohnwu** [comment]
`setns` unfortunately does not have a platform native JNI binding, so apps would have to create the binding itself in native code. But this project is using JNI already, so I think @zhanghai can certainly do this in its root daemon.

**Comment by zhanghai** [comment]
> Apps should not rely on su to switch mount namespaces.

I disagree with this assessment. I think apps can freely say they only want to support what libsu supports - if a su implementation behaves differently then it's an unsupported su implementation. For now both Magisk and SuperSU supports --mount-master.

**Comment by ildar** [comment]
It appears that the Phh's SU simply doesn't need `--mount-master`.
Neo-Backup developers fixed their app to work with it. Some research's made.
See https://github.com/NeoApplications/Neo-Backup/issues/562
Fixing MaterialFiles should be very easy.

**Comment by ildar** [comment]
I see another su/root issue is solved. Anyone wants to look again at this
issue?
